lib/adapters/websockex.ex

```diff
@@ -4,93 +4,101 @@ if Code.ensure_loaded?(WebSockex) do
+                {:ok, ws}
+            after
+              timeout ->
+                # close connection no matter if it is still connecting or not
+                disconnect(ws)
+                {:error, :connection_timeout}
+            end
+
+          {:error, _} = error ->
+            error
+        end
+      end
 
-    @impl Adapter
-    def send(client, payload) do
-      # WebSockex does not support iodata as payload
-      payload = IO.iodata_to_binary(payload)
+      @impl true
+      def handle_connect(connection, %{notify_on_connect: pid} = state) do
+        Adapter.notify_status(pid, {:connected, connection})
+        {:ok, state}
+      end
 
-      try do
-        case WebSockex.send_frame(client, {:text, payload}) do
-          :ok -> :ok
-          {:error, _reason} = error -> error
-        end
-      rescue
-        _ -> {:error, :connection_down}
+      @impl true
+      def handle_disconnect(connection_status, %{on_disconnect: on_disconnect} = state) do
+        on_disconnect.(connection_status)
+        {:ok, state}
       end
-    end
 
-    @impl Adapter
-    def disconnect(client) do
-      Kernel.send(client, :close)
-    end
+      @impl true
+      def handle_frame({_type, frame}, %{on_receive: on_receive} = state) do
+        on_receive.(frame)
+        {:ok, state}
+      end
 
-    defp start_link(url, timeout, args) do
-      %{extra_headers: extra_headers} = args
-
-      opts = [extra_headers: extra_headers]
-
-      case WebSockex.start_link(url, __MODULE__, args, opts) do
-        {:ok, ws} ->
-          # process have started but connection may still not be established
-          # therefore wait for response from handle_connect callback
-          receive do
-            {:connected, _connection} ->
-              {:ok, ws}
-          after
-            timeout ->
-              # close connection no matter if it is still connecting or not
-              close(ws)
-              {:error, :connection_timeout}
-          end
+      @impl true
+      def handle_info(:close, state) do
+        {:close, state}
+      end
 
-        {:error, _} = error ->
-          error
+      def disconnect(connection) do
+        Kernel.send(connection, :close)
       end
     end
 
-    @impl WebSockex
-    def handle_connect(connection, %{notify_on_connect: pid} = state) do
-      notify_status(pid, {:connected, connection})
-      {:ok, state}
-    end
+    @impl Adapter
+    def connect(url, receiver, opts) do
+      timeout = opts[:timeout] || 5000
+      extra_headers = opts[:extra_headers] || []
 
-    @impl WebSockex
-    def handle_disconnect(connection_status, %{message_receiver: message_receiver} = state) do
-      notify_status(message_receiver, {:disconnected, connection_status})
-      {:ok, state}
-    end
+      args = %{
+        on_receive: fn frame -> forward_frame(receiver, frame) end,
+        on_disconnect: fn reason -> notify_status(receiver, {:disconnected, reason}) end,
```

Comment:
I think you overcomplicated that a bit - there's no need to make this inner module so generic - treat it as private for this module and put those calls inside it (you can pass just a receiver)